src/ontology/components/mappings.owl

```diff
@@ -39,1071 +58,5898 @@
     <rdf:Description rdf:about="http://purl.obolibrary.org/obo/UBERON_0000005">
         <oboInOwl:hasDbXref>FBbt:00005157</oboInOwl:hasDbXref>
     </rdf:Description>
+    <owl:Axiom>
+        <owl:annotatedSource rdf:resource="http://purl.obolibrary.org/obo/UBERON_0000005"/>
+        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasDbXref"/>
+        <owl:annotatedTarget>FBbt:00005157</owl:annotatedTarget>
+        <sssom:mapping_justification rdf:resource="https://w3id.org/semapv/vocab/ManualMappingCuration"/>
+    </owl:Axiom>
     <rdf:Description rdf:about="http://purl.obolibrary.org/obo/UBERON_0000010">
         <oboInOwl:hasDbXref>FBbt:00005098</oboInOwl:hasDbXref>
     </rdf:Description>
+    <owl:Axiom>
+        <owl:annotatedSource rdf:resource="http://purl.obolibrary.org/obo/UBERON_0000010"/>
+        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasDbXref"/>
+        <owl:annotatedTarget>FBbt:00005098</owl:annotatedTarget>
+        <sssom:mapping_justification rdf:resource="https://w3id.org/semapv/vocab/ManualMappingCuration"/>
+    </owl:Axiom>
+    <rdf:Description rdf:about="http://purl.obolibrary.org/obo/UBERON_0000016">
+        <oboInOwl:hasDbXref>https://meshb.nlm.nih.gov/record/ui?ui=D007515</oboInOwl:hasDbXref>
```

Comment:
this doesn't seem right

Reply:
It’s not written as a CURIE because ROBOT is not aware of the `https://meshb.nlm.nih.gov/record/ui?=` URL prefix, so it doesn’t know how to condense  such URLs.

We need to add `--add-prefix "MESH: https://meshb.nlm.nih.gov/record/ui?ui=` to the ROBOT call if we want such URLs to be written as CURIEs.

Reply:
This is the only non-curie xref, I would strongly suggest to add the prefix. 

But I am a bit surprised what ROBOT has to do with this, as hasDbXref are not of type "IRI" but of type string; I assume your plugin has to compress? In this case, shouldn't it use the prefixmap supplied by the sssom file (`biomappings.sssom.yml`)?

Reply:
> shouldn't it use the prefixmap supplied by the sssom file (biomappings.sssom.yml)?

It can’t. The prefix map is _not_ part of the SSSOM data model (maybe it should, but it is not). Once the SSSOM file has been read, the code downstream of the SSSOM reader has no access to whatever the original prefix map was.

Reply:
Hmmm. This does not seem ideal. The fact that the `curie_map` is not mentioned in the sssom data model needs to be remedied then, as sssom-py for example just assumes that it is part for the "format" without being part of the "model" - I guess this prevents it from being "passed downstream" according to your thinking.

So internally, how does sssom-java keep track of a `subject_id`? just a URI?

Reply:
> So internally, how does sssom-java keep track of a subject_id? just a URI?

Yes. The prefix map provided in the SSSOM file is used internally by the reader to expand all CURIEs on the fly. Once the reader is done, the client only gets full-length IRIs. This is the only sane way to deal with CURIEs. Expand them as soon as possible (just after reading) and condense them as late as possible (just before writing).

And in this particular situation here, where we generate the `mappings.owl` component from the mappings, I see no reason why the original prefix map should be used, even if the model gave us access to it. Doing otherwise would only open a can of worms, for example: we are merging several mapping sets here (the one from Biomappings is only one among others), they all come with their own prefix maps – how would we solve any conflict between two prefix maps? Should the first entry take precedence? The last entry? Why?

We don’t have to worry about that because all “input” prefix maps are discarded after use. If a prefix map has to be used for writing, it has to be an explicitly specified one, at the discretion of the user.

Reply:
>  sssom-py for example just assumes that it is part for the "format" without being part of the "model"

Yes, that’s exactly the case. The _format_ has a CURIE map, but the _mapping set object_ does not. The CURIE map is only an internal detail of the format. Once the reader has done its job of reading the file, it gives you a _mapping set object_, with all the fields defined in the spec – and nothing more, so the map is _not_ part of it.

Reply:
> Doing otherwise would only open a can of worms, for example: we are merging several mapping sets here (the one from Biomappings is only one among others), they all come with their own prefix maps – how would we solve any conflict between two prefix maps? Should the first entry take precedence? The last entry? Why?

We (in the sense of @cthoyt) have started developing some very well documented processes for prefix standardisation, so its not entirely wild west, but I do agree that there are some kind of conflicts that are hard to overcome. For an ontology project I guess you could argue that mapping sets go in, get expanded, but then get contracted again based on an internally managed prefixmap like https://bioregistry.io/context/obo, this is what we do in sssom-py. Its not enough to use the prefixmaps to go to URI syntax, because different sources will declare different URI expansions for the same exact prefix. 

I still maintain that we should have a prefix standardisation process in place which (roughly) does this:

1. expand CURIEs to URIs according to provided prefixmap (what you do)
2. _always_ compress URIs to CURIEs when writing xrefs, and failing if contraction fails due to missing prefix
3. using https://bioregistry.io/context/obo as the standard map for compression, and add any missing (I doubt there will be many) URIs prefixes in there (its a super fast turnaround). A compromise here would be to keep a small "internal" extension map in the repo to quickly add unsupported uri prefixes to that, and chain this internal prefixmap with the obo one.

Reply:
> For an ontology project I guess you could argue that mapping sets go in, get expanded, but then get contracted again based on an internally managed prefixmap like https://bioregistry.io/context/obo, this is what we do in sssom-py

“Get contracted again” when? Just before writing? Just after reading?

> Its not enough to use the prefixmaps to go to URI syntax

What the heck is the prefix map good for then?

> because different sources will declare different URI expansions for the same exact prefix.

If one source expands `mesh:D123456` to `https://meshb.nlm.nih.gov/record/ui?ui=D123456` and another expands `mesh:D123456` to `http://id.nlm.nih.gov/mesh/D123456`, then they are not referring to the same entities. Maybe they _wanted_ to refer to the same entities, but that’s their problem to fix, it’s out of scope for SSSOM. As I said: garbage in, garbage out.

> 1. expand CURIEs to URIs according to provided prefixmap (what you do)

OK.

> 2. always compress URIs to CURIEs when writing xrefs, and failing if contraction fails due to missing prefix

OK. `sssom-inject` does not fail when a prefix is missing (it writes the uncontracted IRI instead), but that can be configurable.

> 3. using https://bioregistry.io/context/obo as the standard map for compression, and add any missing

OK, so that’s what I said above: any “input” prefix map is discarded, and if contraction is wanted upon writing then an “output” prefix map has to be explicitly provided (which can be done with prefix declarations in SSSOM/T). From the implementation’s point of view it doesn’t matter whether that map is the one from `https://bioregistry.io/context/obo` or any other.

Reply:
> If one source expands mesh:D123456 to https://meshb.nlm.nih.gov/record/ui?ui=D123456 and another expands mesh:D123456 to http://id.nlm.nih.gov/mesh/D123456, then they are not referring to the same entities

I will respond to the rest tomorrow, but this sentence here ⬆️ is at least debatable. We had long years of discussions on this, and the main problem is that there are quite a few resources that do not have a standardised URI expansion  which led to the wild west out there. The main point of bioregistry (from our POV) is to group synonymous expansions into the same "identifier spaces", so they can be standardised. So in fact, `https://meshb.nlm.nih.gov/record/ui?ui=D123456` and `http://id.nlm.nih.gov/mesh/D123456` are the exact same thing. Or rather, to be very precise, insisting on them being _not_ the same thing because they are not the same URI is impractical for most, of not all, use cases. There is huge disagreement on a lot of this, but well, I will just push the practical view over the "formally correct" one.

Reply:
> quite a few resources that do not have a standardised URI expansion which led to the wild west out there. The main point of bioregistry (from our POV) is to group synonymous expansions into the same "identifier spaces", so they can be standardised

That’s the main point of bioregistry, OK. Is that the main point, or even _a_ point, of SSSOM? Is it up to a SSSOM implementation to fix that mess?

Before you answer “yes”, please go read again the SSSOM specification: there is no mention of that problem _anywhere_. I defy anyone to read the spec and to conclude that they must be prepared to deal with the fact that maybe, a subject ID of `https://meshb.nlm.nih.gov/record/ui?ui=D123456` in one mapping could refer to the same entity as `http://id.nlm.nih.gov/mesh/D123456` in another mapping, and that therefore you may need to “reconcile” IDs before doing anything.

If you want(ed) all implementations to implement reconciliation, you should have made that clear, sorry. I wrote SSSOM-Java without that particular problem in mind, and changing that now would basically mean having to rewrite the entire thing from scratch. Not going to happen. Reconciliation of messy IDs is forever going to be out of scope for SSSOM-Java.

Reconciliation can always be performed upstream of SSSOM-Java if/when needed.

Reply:
> insisting on them being not the same thing because they are not the same URI is impractical for most, of not all, use cases. 

It is practical enough for all the use cases I care about. For example, it was practical enough to rebuild the bridge pipeline in Uberon, which is the primary reason why I wrote SSSOM-Java in the first place.

That it is impractical for the use-case of creating cross-references from the Biomappings set is the least of my concerns.

Reply:
By the way:

> but then get contracted again based on an internally managed prefixmap like https://bioregistry.io/context/obo, **this is what we do in sssom-py**

Is that really what you do? Because parsing the Biomappings set with `sssom-py` (last version, 0.4.2) still yields `mesh` CURIEs that expands to `https://meshb.nlm.nih.gov/record/ui?ui=`, despite the OBO context supposedly associating MESH identifiers to the `http://id.nlm.nih.gov/mesh/` prefix.

Reply:
> That’s the main point of bioregistry, OK. Is that the main point, or even a point, of SSSOM? Is it up to a SSSOM implementation to fix that mess?
Before you answer “yes”, please go read again the SSSOM specification: there is no mention of that problem anywhere. I defy anyone to read the spec and to conclude that they must be prepared to deal with the fact that maybe, a subject ID of https://meshb.nlm.nih.gov/record/ui?ui=D123456 in one mapping could refer to the same entity as http://id.nlm.nih.gov/mesh/D123456 in another mapping, and that therefore you may need to “reconcile” IDs before doing anything.

Sorry I didn't mean to say this "should be solved by SSSOM" - it is, as you say, out of scope! The argument here is  on the level of the ETL pipeline (Uberon), and here I think it is justifiable to say that we reconcile prefixes before republishing them for our users.

> If you want(ed) all implementations to implement reconciliation, you should have made that clear, sorry. I wrote SSSOM-Java without that particular problem in mind, and changing that now would basically mean having to rewrite the entire thing from scratch. Not going to happen. Reconciliation of messy IDs is forever going to be out of scope for SSSOM-Java.

No, reconciliation should be a special purpose convenience method, not a part of the spec! I never meant to suggest that.

> It is practical enough for all the use cases I care about. For example, it was practical enough to rebuild the bridge pipeline in Uberon, which is the primary reason why I wrote SSSOM-Java in the first place.

In my understanding of the problem the bridge file generation is a bit unrelated to the question of whether "two URIs can refer to the same entity". This PR here is about redistributing xrefs, not the bridge pipeline, right? So the argument is purely about standardising whatever we redistribute, and again, yes, this happens downstream (or ideally upstream, but certainly offstream) of SSSOM.

> ..parsing the Biomappings set with sssom-py (last version, 0.4.2) still yields mesh CURIEs that expands to https://meshb.nlm.nih.gov/record/ui?ui=, despite the OBO context supposedly associating MESH identifiers to the http://id.nlm.nih.gov/mesh/ prefix.

This is a mistake in the prefixmap, I will fix it. (The prefix map is unfortunately an eternal work in progress).

Reply:
> Sorry I didn't mean to say this "should be solved by SSSOM" - it is, as you say, out of scope! […] reconciliation should be a special purpose convenience method

So in the scope of SSSOM and unless you’re running in a special mode in which you explicitly want to mangle identifiers, if two identical CURIEs expand to different full-length identifiers (because of a different prefix map), the identifiers are _not_ the same and should not be treated as if they were the same. Good.

>  I think it is justifiable to say that we reconcile prefixes before republishing them for our users

That’s debatable (who are we to say which URL prefix is to be used? for example, the Allen institute now uses `https://purl.brain-bican.org/ontology/xxxo/XXX_` as its prefix for its brain ontologies, why would we say that “no no no, we know better than the folks at the Allen, the correct prefix is `http://purl.obolibrary.org/obo/XXX_`”?), but okay, if it’s an explicit move and not something done silently by the implementation without any user opt-in. I can implement  such a “reconciliation” as an option to `sssom-cli`.


> In my understanding of the problem the bridge file generation is a bit unrelated to the question of whether "two URIs can refer to the same entity".

That question is _very_ relevant for the generation of the bridges. The bridges are OWL files, and in OWL (as in RDF), only full-length identifiers matter, and they are compared as opaque strings without any notion of “URL prefix“. Maybe outside of OWL/RDF you can pretend that `https://meshb.nlm.nih.gov/record/ui?ui=D123456` in one file and `http://id.nlm.nih.gov/mesh/D123456` in another file are actually the same thing (because magic), but `<owl:Class rdf:about="https://meshb.nlm.nih.gov/record/ui?ui=D123456">` and `<owl:Class rdf:about="http://id.nlm.nih.gov/mesh/D123456">` are _never_ going to be considered the same, unless you have an explicit equivalence axiom.

So it is very important to be sure we are dealing with the correct URL prefix when we generate the bridges, if you don’t want to generate a bridge that connect to a different island than the one intended. Right now, we can be sure of that because we only generate the bridges from two kinds of mappings:

*) mappings that we (Uberon editors) maintain ourselves in Uberon: we maintain them, so we can check that we are using the correct prefixes;
*) mappings from CL, FBbt, and ZFA: we don’t have control over them (well, I control the FBbt ones but let’s assume I don’t) but they are mappings between OBO resources only (CL-to-Uberon, FBbt-to-Uberon, ZFA-to-Uberon), so it’s all `http://purl.obolibrary.org/obo` and there is little risk of anything fishy happening with the identifiers (apart from an always possible silly mistake).

> This PR here is about redistributing xrefs, not the bridge pipeline, right?

This PR, yes. This discussion, no. For distributing xrefs, this entire discussion is actually irrelevant, since the xrefs are pure CURIEs whose interpretation is entirely left to the user, there is no associated prefix name declaration. Which makes those cross-references much less useful than they could be, but well, that’s what they have always been – we are generating and distributing them only for backwards compatibility for the sake of people using tools that know nothing else than `oboInOwl:hasDbXref`.

The only real problem in this PR was the fact that the `https://meshb.nlm.nih.gov/record/ui?ui=` was not declared, which prevented the `sssom-inject` command from shortening `https://meshb.nlm.nih.gov/record/ui?ui=D123456` into `MESH:D123456`. This is now fixed and an option has been added to raise an error if an undeclared prefix sneaks in again.

Everything else is a more general discussion that could (should?) have happened elsewhere.